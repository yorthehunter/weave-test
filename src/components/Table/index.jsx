import React from 'react';
import { array, arrayOf, oneOf, string } from 'prop-types';
import shortid from 'shortid';
import CustomTag from '../utilities/CustomTag';

const componentPrefix = 'weave-table';

const createTDs = (tdData, isHeading) => {
  const Tag = isHeading ? 'th' : 'td';

  return (
    Array.isArray(tdData) ?
      tdData.map((td) => (
        <Tag
          key={shortid.generate()}
          dangerouslySetInnerHTML={{ __html: td }}
        />
      ))
      :
      <Tag dangerouslySetInnerHTML={{ __html: tdData }} />
  );
};

const createTRs = (trData) => (
  trData.map((tr) => (
    <tr key={shortid.generate()}>
      {createTDs(tr)}
    </tr>
  ))
);

const VerticalTable = ({ headings, data }) => (
  <React.Fragment>
    <thead>
      <tr>
        {createTDs(headings, true)}
      </tr>
    </thead>
    <tbody>
      {createTRs(data)}
    </tbody>
  </React.Fragment>
);

const HorizontalTable = ({ headings, data }) => (
  <tbody>
    {
      headings.map((tr, index) => (
        <tr key={shortid.generate()}>
          <th
            style={{ textAlign: 'right' }}
            scope="row"
          >
            {headings[index]}
          </th>
          {createTDs(data[index])}
        </tr>
      ))
    }
  </tbody>
);

const Table = (props) => {
  const {
    type,
    orientation,
    size,
  } = props;

  const tableType = (modifier) => {
    if (Array.isArray(modifier)) {
      const modifiers = modifier.map((item) => `${componentPrefix}--${item}`);
      return modifiers;
    } else if (modifier) {
      return `${componentPrefix}--${modifier}`;
    } else {
      return null;
    }
  };

  const componentClasses = [
    componentPrefix,
    tableType(type),
  ];

  return (
    <CustomTag
      {...props}
      tag="table"
      componentClasses={componentClasses}
    >
      {
        orientation === 'horizontal' ?
          <HorizontalTable {...props} />
        :
          <VerticalTable {...props} />
      }
    </CustomTag>
  );
};

Table.defaultProps = {
  orientation: 'vertical',
  type: null,
  size: null,
};
Table.propTypes = {
  orientation: oneOf(['vertical', 'horizontal']),
  type: string,
  size: string,
};
const commonDefaultProps = {
  headings: null,
};
const commonPropTypes = {
  headings: arrayOf(string),
  data: array.isRequired,
};
VerticalTable.defaultProps = commonDefaultProps;
VerticalTable.propTypes = commonPropTypes;
HorizontalTable.defaultProps = commonDefaultProps;
HorizontalTable.propTypes = commonPropTypes;

export default Table;
