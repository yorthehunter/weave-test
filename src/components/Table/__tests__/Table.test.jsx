import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Table from '../';

describe('Table', () => {
  const props = {
    data: [
      'one thing',
      'two things',
      'three things',
      'four',
    ],
    headings: [
      'heading one',
      'heading two',
      'heading three',
      'heading four',
    ],
  };

  it('correctly renders a vertical table', () => {
    const wrapper = render(<Table {...props} />);
    const rendered = toJson(wrapper);
    expect(rendered).toMatchSnapshot();
  });

  it('correctly renders a horizontal table', () => {
    const wrapper = render(<Table {...props} orientation="horizontal" />);
    const rendered = toJson(wrapper);
    expect(rendered).toMatchSnapshot();
  });

  it('renders a basic table when passed a basic type', () => {
    const wrapper = render(<Table {...props} type="basic" />);
    expect(wrapper.find('.weave-table--basic').hasClass('weave-table--striped')).toEqual(false);
  });
});
