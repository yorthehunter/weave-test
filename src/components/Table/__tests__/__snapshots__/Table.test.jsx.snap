// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table correctly renders a horizontal table 1`] = `
<table
  class="weave-table weave-table--basic weave-table--bordered weave-table--striped weave-table--full-width weave-table--hoverable"
>
  <tbody>
    <tr>
      <th
        scope="row"
        style="text-align:right"
      >
        heading one
      </th>
      <td>
        one thing
      </td>
    </tr>
    <tr>
      <th
        scope="row"
        style="text-align:right"
      >
        heading two
      </th>
      <td>
        two things
      </td>
    </tr>
    <tr>
      <th
        scope="row"
        style="text-align:right"
      >
        heading three
      </th>
      <td>
        three things
      </td>
    </tr>
    <tr>
      <th
        scope="row"
        style="text-align:right"
      >
        heading four
      </th>
      <td>
        four
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table correctly renders a vertical table 1`] = `
<table
  class="weave-table weave-table--basic weave-table--bordered weave-table--striped weave-table--full-width weave-table--hoverable"
>
  <thead>
    <tr>
      <th>
        heading one
      </th>
      <th>
        heading two
      </th>
      <th>
        heading three
      </th>
      <th>
        heading four
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        one thing
      </td>
    </tr>
    <tr>
      <td>
        two things
      </td>
    </tr>
    <tr>
      <td>
        three things
      </td>
    </tr>
    <tr>
      <td>
        four
      </td>
    </tr>
  </tbody>
</table>
`;
